src/components/calendar-with-shortcuts/CalendarWithShortcuts.tsx

```diff
@@ -0,0 +1,132 @@
+  );
+
+  const onCalendarChange = useCallback(
+    (val) => {
+      updateSelectedDates(val);
+      if (props.onChange) {
+        props.onChange(val);
+      }
+    },
+    [props, updateSelectedDates],
+  );
+
+  const updateDateView = useCallback(
+    (date: Date | undefined) => {
+      if (isValid(date)) {
+        setViewDate(getAsNewDate(date));
+      }
+    },
+    [setViewDate],
+  );
+
+  const goToToday = useCallback(() => updateDateView(new Date()), [updateDateView]);
+  const goToRangeStart = useCallback(() => updateDateView(rangeStart), [rangeStart, updateDateView]);
+  const goToRangeEnd = useCallback(() => updateDateView(rangeEnd), [rangeEnd, updateDateView]);
+  const toggleDate = useCallback(() => {
+    let updateVal;
+    if (props.isMultiSelector && multiDates) {
+      const values = Object.values(multiDates).sort((a, b) =>
+        isValid(a) && isValid(b) ? a.getTime() - b.getTime() : 0,
+      );
+      updateVal = values[toggleDateIndex];
+      setToggleDateIndex(toggleDateIndex < values.length - 1 ? toggleDateIndex + 1 : 0);
+    } else if (isValid(selectedDate)) {
+      updateVal = selectedDate;
+    }
+    updateDateView(updateVal);
+  }, [props.isMultiSelector, multiDates, selectedDate, toggleDateIndex, updateDateView]);
+
+  const defaultShortcutButtons: Array<ShortcutButtonModel> = useMemo(
+    () => [
+      {
+        buttonText: 'Today',
+        onButtonClick: goToToday,
+      },
+      {
+        buttonText: 'Selected Date',
+        viewTypes: ['Normal', 'Multiple'],
+        onButtonClick: toggleDate,
+      },
+      {
+        buttonText: 'Range Start',
+        viewTypes: ['Range'],
+        onButtonClick: goToRangeStart,
+      },
+      {
+        buttonText: 'Range End',
+        viewTypes: ['Range'],
+        onButtonClick: goToRangeEnd,
+      },
+    ],
+    [goToRangeEnd, goToRangeStart, goToToday, toggleDate],
+  );
+
+  const viewType = props.isMultiSelector ? 'Multiple' : props.isRangeSelector ? 'Range' : 'Normal';
+
+  const shortcutButtonsToShow = props.showDefaultShortcuts
+    ? props.shortcutButtons
+      ? defaultShortcutButtons.concat(props.shortcutButtons)
+      : defaultShortcutButtons
+    : props.shortcutButtons;
+
+  useEffect(
+    () => updateSelectedDates(props.value),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [],
+  );
+
+  const commonProps = useMemo<CalendarProps>(
+    () => ({ ...props, viewDate: viewDate, onChange: onCalendarChange }),
+    [onCalendarChange, props, viewDate],
+  );
+
+  return (
+    <div ref={ref} style={styles.root} className="rc_shortcut_cal_root">
+      <ShortcutBar shortcutButtons={shortcutButtonsToShow} viewType={viewType} updateView={updateDateView} />
+      <Calendar {...commonProps} />
+    </div>
+  );
+}
+
+// export const CalendarWithShortcuts = memo(CalendarWithShortcutsComponent);
+const CalendarWithShortcuts = React.forwardRef(CalendarWithShortcutsRef);
+
```

Comment:
Not sure if ref is needed here, let me know if this can be a regular component :)